File: backend/internal/shared/topology/walker.go
// Package topology is the Go mirror of app/lib/shared/topology/walker.ts.
//
// PLA-0044: single unified flatten of a parent_id-linked node forest with
// visibility, collapse, orphan-drop, cycle-guard, and sibling-spine
// encoding. Cross-runtime parity is enforced by golden fixtures at
// dev/fixtures/shared/topology — see walker_test.go which loads the
// same JSON files Vitest does and asserts byte-identical row projections.
//
// Catalogued in docs/c_shared_methods.md.
package topology

import "sort"

// Node is the minimum required shape. Callers wrap their domain type
// (e.g. orgdesign.Node, orgdesign.MyGrant) in an IDOf/ParentOf accessor
// pair so the walker can stay free of generics-on-fields.
type Node interface {
	GetID() string
	GetParentID() *string
}

// FlattenedRow mirrors app/lib/shared/topology/walker.ts FlattenedRow.
// ID is denormalised onto the row to make JSON-fixture comparison
// trivial across runtimes — callers can still reach the underlying
// Node via the typed walker wrappers below.
type FlattenedRow struct {
	ID                   string `json:"id"`
	Depth                int    `json:"depth"`
	HasChildren          bool   `json:"hasChildren"`
	Collapsed            bool   `json:"collapsed"`
	IsFirst              bool   `json:"isFirst"`
	IsLast               bool   `json:"isLast"`
	HasVisibleChildren   bool   `json:"hasVisibleChildren"`
	AncestorMoreChildren []bool `json:"ancestorMoreChildren"`
}

// Edge is an abstract parent→child pair. Coordinates are attached
// downstream by dagre / d3-zoom in the frontend.
type Edge struct {
	Source string `json:"source"`
	Target string `json:"target"`
}

// Opts mirrors WalkOpts in TS.
//
// Collapsed: ids that emit their own row but suppress their subtree.
// Less: sibling sort comparator (sort.Slice convention — true means a
// comes before b). Filter: pre-walk predicate; returning false drops
// the node and its subtree. MaxDepth: cycle guard, default 12.
type Opts[T Node] struct {
	Collapsed map[string]struct{}
	Less      func(a, b T) bool
	Filter    func(n T) bool
	MaxDepth  int
}

// Result mirrors WalkResult in TS, minus ChildrenOf — callers that need
// it can construct it from the same sorted buckets via BuildChildrenOf.
type Result[T Node] struct {
	Rows         []FlattenedRow
	VisibleIDs   map[string]struct{}
	VisibleEdges []Edge
	ChildrenOf   map[string][]T // empty string key = roots (mirrors TS null)
}

const defaultMaxDepth = 12

// Walk traverses nodes and emits the same shape as the TS walker. See
// the package comment for parity guarantees.
func Walk[T Node](nodes []T, opts Opts[T]) Result[T] {
	maxDepth := opts.MaxDepth
	if maxDepth == 0 {
		maxDepth = defaultMaxDepth
	}

	// Pass 1: filter + index by id.
	byID := make(map[string]T, len(nodes))
	for _, n := range nodes {
		if opts.Filter != nil && !opts.Filter(n) {
			continue
		}
		byID[n.GetID()] = n
	}

	// Pass 2: bucket children. Orphan policy = drop (matches TS).
	childrenOf := make(map[string][]T)
	for _, n := range byID {
		pid := n.GetParentID()
		var key string
		if pid == nil {
			key = "" // root sentinel — matches TS null
		} else if _, ok := byID[*pid]; ok {
			key = *pid
		} else {
			continue // orphan — drop
		}
		childrenOf[key] = append(childrenOf[key], n)
	}
	for k, bucket := range childrenOf {
		b := bucket
		sort.SliceStable(b, func(i, j int) bool { return opts.Less(b[i], b[j]) })
		childrenOf[k] = b
	}

	// Pass 3: depth-first emission.
	rows := make([]FlattenedRow, 0, len(byID))
	visibleIDs := make(map[string]struct{}, len(byID))
	visibleEdges := make([]Edge, 0)

	var walk func(parentID string, depth int, pathMoreChildren []bool)
	walk = func(parentID string, depth int, pathMoreChildren []bool) {
		if depth > maxDepth {
			return
		}
		kids := childrenOf[parentID]
		for idx, node := range kids {
			id := node.GetID()
			childKids := childrenOf[id]
			hasChildren := len(childKids) > 0
			_, isCollapsed := opts.Collapsed[id]
			isFirst := idx == 0
			isLast := idx == len(kids)-1
			hasVisibleChildren := hasChildren && !isCollapsed
			visibleIDs[id] = struct{}{}
			if parentID != "" {
				visibleEdges = append(visibleEdges, Edge{Source: parentID, Target: id})
			}
			// Snapshot the path slice into a fresh backing array so
			// later appends in sibling subtrees don't mutate this
			// row's value. JSON-encoded as a stable []bool.
			amc := make([]bool, len(pathMoreChildren))
			copy(amc, pathMoreChildren)
			rows = append(rows, FlattenedRow{
				ID:                   id,
				Depth:                depth,
				HasChildren:          hasChildren,
				Collapsed:            isCollapsed,
				IsFirst:              isFirst,
				IsLast:               isLast,
				HasVisibleChildren:   hasVisibleChildren,
				AncestorMoreChildren: amc,
			})
			if hasVisibleChildren {
				childPath := append(amc[:len(amc):len(amc)], !isLast)
				walk(id, depth+1, childPath)
			}
		}
	}
	walk("", 0, []bool{})

	return Result[T]{
		Rows:         rows,
		VisibleIDs:   visibleIDs,
		VisibleEdges: visibleEdges,
		ChildrenOf:   childrenOf,
	}
}

File: app/lib/shared/topology/walker.ts
// PLA-0044: unified topology-traversal engine.
//
// Single generic walker used by every consumer that needs to flatten a
// parent_id-linked node forest into an ordered, visibility-aware row list:
// canvas dagre layout, tree-state hook, topology flyout, scope rail, BFF
// scope-clamped tree responses. The Go mirror at
// backend/internal/shared/topology lives behind a golden-fixture parity
// test (dev/fixtures/shared/topology) so both runtimes produce
// byte-identical output for the same input.
//
// Catalogued in docs/c_shared_methods.md.

export type TopologyNode = {
  id: string;
  parent_id: string | null;
};

export type FlattenedRow<T extends TopologyNode> = {
  node: T;
  depth: number;
  hasChildren: boolean;
  collapsed: boolean;
  isFirst: boolean;
  isLast: boolean;
  hasVisibleChildren: boolean;
  // Tree-spine encoding. Length = depth. Index d entry encodes "does the
  // ancestor at depth d have a later sibling subtree below this row?" —
  // renderers paint a vertical at column d*STEP + STEP/2 through the row
  // wherever the entry is true. The entry for the immediate parent
  // (index depth-1) is included; consumers that draw their own elbow on
  // that column can ignore it. Depth 0 rows always get an empty array.
  ancestorMoreChildren: boolean[];
};

export type WalkOpts<T extends TopologyNode> = {
  // Collapsed node ids. A collapsed node still emits its own row; its
  // subtree is skipped and not counted toward visibleEdges.
  collapsed: Set<string>;
  // Sibling order comparator. Applied to children of each parent before
  // emission. Caller injects byPosition / byLabel / etc.
  sort: (a: T, b: T) => number;
  // Pre-filter. Returning false drops the node AND its subtree (e.g.
  // archived_at !== null). Default = keep all.
  filter?: (n: T) => boolean;
  // Hard recursion cap, matches DescendantNodeIDs (12). On overflow the
  // walker stops descending; no synthetic row is emitted.
  maxDepth?: number;
};

export type WalkResult<T extends TopologyNode> = {
  rows: FlattenedRow<T>[];
  visibleIds: Set<string>;
  visibleEdges: Array<{ source: string; target: string }>;
  childrenOf: Map<string | null, T[]>;
};

const DEFAULT_MAX_DEPTH = 12;

export function walkTopology<T extends TopologyNode>(
  nodes: T[],
  opts: WalkOpts<T>,
): WalkResult<T> {
  const { collapsed, sort, filter, maxDepth = DEFAULT_MAX_DEPTH } = opts;

  // Pass 1: build childrenOf, applying filter. Orphan policy = drop:
  // a node whose parent_id is set but whose parent is missing/filtered
  // is excluded entirely (it does NOT get re-rooted under null). This
  // was the ScopeRail "phantom D" bug.
  const byId = new Map<string, T>();
  for (const n of nodes) {
    if (filter && !filter(n)) continue;
    byId.set(n.id, n);
  }

  const childrenOf = new Map<string | null, T[]>();
  for (const n of byId.values()) {
    const key =
      n.parent_id === null
        ? null
        : byId.has(n.parent_id)
          ? n.parent_id
          : undefined; // sentinel: drop
    if (key === undefined) continue;
    let bucket = childrenOf.get(key);
    if (!bucket) {
      bucket = [];
      childrenOf.set(key, bucket);
    }
    bucket.push(n);
  }
  for (const bucket of childrenOf.values()) bucket.sort(sort);

  // Pass 2: depth-first emission with visibility tracking.
  const rows: FlattenedRow<T>[] = [];
  const visibleIds = new Set<string>();
  const visibleEdges: Array<{ source: string; target: string }> = [];

  const walk = (
    parentId: string | null,
    depth: number,
    pathMoreChildren: boolean[],
  ): void => {
    if (depth > maxDepth) return;
    const kids = childrenOf.get(parentId) ?? [];
    kids.forEach((node, idx) => {
      const childKids = childrenOf.get(node.id) ?? [];
      const hasChildren = childKids.length > 0;
      const isCollapsed = collapsed.has(node.id);
      const isFirst = idx === 0;
      const isLast = idx === kids.length - 1;
      const hasVisibleChildren = hasChildren && !isCollapsed;
      visibleIds.add(node.id);
      if (parentId !== null) {
        visibleEdges.push({ source: parentId, target: node.id });
      }
      rows.push({
        node,
        depth,
        hasChildren,
        collapsed: isCollapsed,
        isFirst,
        isLast,
        hasVisibleChildren,
        ancestorMoreChildren: pathMoreChildren,
      });
      if (hasVisibleChildren) {
        // Children at depth+1 carry a path array of length depth+1.
        // The new entry (index depth) records "does the ancestor at
        // depth `depth` — i.e. this row — have a later sibling
        // subtree?" → !isLast. Renderers that suppress the depth-0
        // spine can ignore index 0; the walker always emits uniformly.
        const childPath = [...pathMoreChildren, !isLast];
        walk(node.id, depth + 1, childPath);
      }
    });
  };
  walk(null, 0, []);

  return { rows, visibleIds, visibleEdges, childrenOf };
}

// Convenience comparators used across consumers. Callers pass these
// directly as opts.sort.

export const byPosition = <
  T extends TopologyNode & { position: number; id: string },
>(
  a: T,
  b: T,
): number => {
  if (a.position !== b.position) return a.position - b.position;
  return a.id.localeCompare(b.id);
};

export const byLabel = <
  T extends TopologyNode & { __label?: string; name?: string },
>(
  a: T,
  b: T,
): number => {
  const la = a.__label ?? a.name ?? a.id;
  const lb = b.__label ?? b.name ?? b.id;
  return la.localeCompare(lb);
};
